### ISSUE TITLE
[MQTT][Linux] Segmentation fault due to looping call to xio_close

### ISSUE THREAD

**Issue by lmussier** [Azure/azure-iot-sdk-c#53]
Hi,

On tag 2017-01-27, we experience a crash du to a looping call to the xio close function. Here is a part of the back trace.
```
Thread 1 "iothubAgentUT" received signal SIGSEGV, Segmentation fault.
0x00000000005c1fd8 in onOpenComplete (context=0x7fffec001220, open_result=<optimized out>) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:374
374	                mqtt_client->fnOnErrorCallBack(mqtt_client, MQTT_CLIENT_CONNECTION_ERROR, mqtt_client->errorCBCtx);
(gdb) bt
#0  0x00000000005c1fd8 in onOpenComplete (context=0x7fffec001220, open_result=<optimized out>) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:374
#1  0x00000000005d514e in indicate_open_complete (ws_io_instance=0x7fffe4000bc0, ws_io_instance=0x7fffe4000bc0, open_result=IO_OPEN_CANCELLED) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/wsio.c:83
#2  wsio_close (ws_io=0x7fffe4000bc0, on_io_close_complete=0x5c1500 <on_connection_closed>, on_io_close_complete_context=0x7fffff7ff098) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/wsio.c:834
#3  0x00000000005c8bb6 in xio_close (xio=<optimized out>, on_io_close_complete=on_io_close_complete@entry=0x5c1500 <on_connection_closed>, callback_context=callback_context@entry=0x7fffff7ff098)
    at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/xio.c:116
#4  0x00000000005c1ffa in close_connection (mqtt_client=0x7fffec001220) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:66
#5  onOpenComplete (context=0x7fffec001220, open_result=<optimized out>) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:376
#6  0x00000000005d514e in indicate_open_complete (ws_io_instance=0x7fffe4000bc0, ws_io_instance=0x7fffe4000bc0, open_result=IO_OPEN_CANCELLED) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/wsio.c:83
#7  wsio_close (ws_io=0x7fffe4000bc0, on_io_close_complete=0x5c1500 <on_connection_closed>, on_io_close_complete_context=0x7fffff7ff138) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/wsio.c:834
#8  0x00000000005c8bb6 in xio_close (xio=<optimized out>, on_io_close_complete=on_io_close_complete@entry=0x5c1500 <on_connection_closed>, callback_context=callback_context@entry=0x7fffff7ff138)
    at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/xio.c:116
#9  0x00000000005c1ffa in close_connection (mqtt_client=0x7fffec001220) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:66
#10 onOpenComplete (context=0x7fffec001220, open_result=<optimized out>) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:376
#11 0x00000000005d514e in indicate_open_complete (ws_io_instance=0x7fffe4000bc0, ws_io_instance=0x7fffe4000bc0, open_result=IO_OPEN_CANCELLED) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/wsio.c:83
#12 wsio_close (ws_io=0x7fffe4000bc0, on_io_close_complete=0x5c1500 <on_connection_closed>, on_io_close_complete_context=0x7fffff7ff1d8) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/wsio.c:834
#13 0x00000000005c8bb6 in xio_close (xio=<optimized out>, on_io_close_complete=on_io_close_complete@entry=0x5c1500 <on_connection_closed>, callback_context=callback_context@entry=0x7fffff7ff1d8)
    at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/xio.c:116
#14 0x00000000005c1ffa in close_connection (mqtt_client=0x7fffec001220) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:66
#15 onOpenComplete (context=0x7fffec001220, open_result=<optimized out>) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:376
#16 0x00000000005d514e in indicate_open_complete (ws_io_instance=0x7fffe4000bc0, ws_io_instance=0x7fffe4000bc0, open_result=IO_OPEN_CANCELLED) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/wsio.c:83
#17 wsio_close (ws_io=0x7fffe4000bc0, on_io_close_complete=0x5c1500 <on_connection_closed>, on_io_close_complete_context=0x7fffff7ff278) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/wsio.c:834
#18 0x00000000005c8bb6 in xio_close (xio=<optimized out>, on_io_close_complete=on_io_close_complete@entry=0x5c1500 <on_connection_closed>, callback_context=callback_context@entry=0x7fffff7ff278)
    at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/xio.c:116
#19 0x00000000005c1ffa in close_connection (mqtt_client=0x7fffec001220) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:66
#20 onOpenComplete (context=0x7fffec001220, open_result=<optimized out>) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:376
#21 0x00000000005d514e in indicate_open_complete (ws_io_instance=0x7fffe4000bc0, ws_io_instance=0x7fffe4000bc0, open_result=IO_OPEN_CANCELLED) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/wsio.c:83
#22 wsio_close (ws_io=0x7fffe4000bc0, on_io_close_complete=0x5c1500 <on_connection_closed>, on_io_close_complete_context=0x7fffff7ff318) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/wsio.c:834
#23 0x00000000005c8bb6 in xio_close (xio=<optimized out>, on_io_close_complete=on_io_close_complete@entry=0x5c1500 <on_connection_closed>, callback_context=callback_context@entry=0x7fffff7ff318)
    at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/xio.c:116
#24 0x00000000005c1ffa in close_connection (mqtt_client=0x7fffec001220) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:66
#25 onOpenComplete (context=0x7fffec001220, open_result=<optimized out>) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:376
#26 0x00000000005d514e in indicate_open_complete (ws_io_instance=0x7fffe4000bc0, ws_io_instance=0x7fffe4000bc0, open_result=IO_OPEN_CANCELLED) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/wsio.c:83
#27 wsio_close (ws_io=0x7fffe4000bc0, on_io_close_complete=0x5c1500 <on_connection_closed>, on_io_close_complete_context=0x7fffff7ff3b8) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/wsio.c:834
#28 0x00000000005c8bb6 in xio_close (xio=<optimized out>, on_io_close_complete=on_io_close_complete@entry=0x5c1500 <on_connection_closed>, callback_context=callback_context@entry=0x7fffff7ff3b8)
    at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/xio.c:116
#29 0x00000000005c1ffa in close_connection (mqtt_client=0x7fffec001220) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:66
#30 onOpenComplete (context=0x7fffec001220, open_result=<optimized out>) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:376
#31 0x00000000005d514e in indicate_open_complete (ws_io_instance=0x7fffe4000bc0, ws_io_instance=0x7fffe4000bc0, open_result=IO_OPEN_CANCELLED) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/wsio.c:83
#32 wsio_close (ws_io=0x7fffe4000bc0, on_io_close_complete=0x5c1500 <on_connection_closed>, on_io_close_complete_context=0x7fffff7ff458) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/wsio.c:834
#33 0x00000000005c8bb6 in xio_close (xio=<optimized out>, on_io_close_complete=on_io_close_complete@entry=0x5c1500 <on_connection_closed>, callback_context=callback_context@entry=0x7fffff7ff458)
    at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/c-utility/src/xio.c:116
#34 0x00000000005c1ffa in close_connection (mqtt_client=0x7fffec001220) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:66
#35 onOpenComplete (context=0x7fffec001220, open_result=<optimized out>) at /home/lmussier/workspace/IoTHubAgent/externals/azure-iot-sdk-c/umqtt/src/mqtt_client.c:376
```
I've avoid this crash with that patch :
```
diff --git a/src/mqtt_client.c b/src/mqtt_client.c
index 44bf153..e4959b2 100644
--- a/src/mqtt_client.c
+++ b/src/mqtt_client.c
@@ -373,7 +373,7 @@ static void onOpenComplete(void* context, IO_OPEN_RESULT open_result)
             {
                 mqtt_client->fnOnErrorCallBack(mqtt_client, MQTT_CLIENT_CONNECTION_ERROR, mqtt_client->errorCBCtx);
             }
-            close_connection(mqtt_client);
+            //close_connection(mqtt_client);
         }
     }
     else
```
But it's not a good way, since close_connection is not called this part of code is not executed, the client is in a weird state
```
    mqtt_client->socketConnected = false;
    mqtt_client->clientConnected = false;
```
I end up with this patch :
```
diff --git a/src/mqtt_client.c b/src/mqtt_client.c
index 44bf153..cb09792 100644
--- a/src/mqtt_client.c
+++ b/src/mqtt_client.c
@@ -373,7 +373,9 @@ static void onOpenComplete(void* context, IO_OPEN_RESULT open_result)
             {
                 mqtt_client->fnOnErrorCallBack(mqtt_client, MQTT_CLIENT_CONNECTION_ERROR, mqtt_client->errorCBCtx);
             }
-            close_connection(mqtt_client);
+            //close_connection(mqtt_client);^M
+            mqtt_client->socketConnected = false;
+            mqtt_client->clientConnected = false;
         }
     }
     else
```
Could you tell me if this is a valid correction?
If yes I'll keep my patch until the bug is fixed.

**Comment by dcristoloveanu** [comment]
Hi @lmussier,

The way I was able to repro this is by having the open (WebSockets connect) timeout when using WebSockets. Is that the way you observed it too?

There are 2 issues:

1.
The issue that you pointed out where I agree that we don't need to close if we're signaled via open_complete of an error or cancelled.

I'll check in your fix as soon as I can and let you know.

-----

2.
There is another issue that the open_complete with CANCELLED is indicated without setting the state to NOT_OPEN. This means that if the consumer calls close in the open_complete callback then the cycle would repeat.

After fixing this, I believe there is a deeper problem here.
wsio triggers callback from the underlying lws callback received from lws. This is an issue because the consumer of wsio could call close which would destroy the lws context, which would result in a crash.

As a note, in parallel we have an effort to replace lws with a more lightweight client only WebSocket implementation that would use our IO abstraction. This work is already started and I wonder if the above workaround that you proposed is good enough for you until we have the lws dependency replaced with the new uws_client we are putting in.

I think fixing this callback issue involves rewriting a lot of the wsio code, but that would already be obsolete by the time we put in the new wsio based on uws_client. Thus I would not make this fix now.

Thanks,
/Dan

**Comment by lmussier** [comment]
Hi @dcristoloveanu,

Point 1/ is crystal clear, I'll grab the fix once it is available.
For point 2, for now since the workaround does not seems to bring draw backs it ok for us. About uws_client doyou have an idea of when it will be available ?

Thanks

**Comment by dcristoloveanu** [comment]
@lmussier 

A fix has been checked in to the master branch for point 1) of this issue.
For point 2) "when will the newer websocket implementation based on uws_client be available?" the answer I have is approx 1 mo (give or take).

I will close this issue since work for 2) is tracked internally and we will announce in the release notes when the new websocket implementation is enabled.

If you still have issues for the point 1) above please reopen and I will have a look.

I hope this helps,
/Dan

**Comment by lmussier** [comment]
I'll check the 1) fix.
Thanks.

**Comment by lmussier** [comment]
@dcristoloveanu I've check with 297dd732588210f2792ccf92c9d92a2ff4235096, it works for me.
Thanks

Ps: I'll let you close if needed, since you close/open this issue.

**Comment by dcristoloveanu** [comment]
@lmussier 
"About uws_client do you have an idea of when it will be available ?"

We have just checked in to master the switch to the newer uws client. 
We'd love any feedback on it. Normally this should be transparent for the users, it is not intended to reflect in apublic API change.

Thanks,
/Dan

**Comment by lmussier** [comment]
@dcristoloveanu sweet !
We'll check that for sure
